### PR TITLE
fix stack_hex not displaying correctly

### LINKED ISSUE #41: stack_hex not displaying correctly when stack contains the number 0
examples:

    script      0 OP_NOT
    script_hex  0091
    stack       1
    stack_hex   01
    results     success

    script      1 OP_NOT
    script_hex  5191
    stack       0
    stack_hex                                {missing 00}
    results     success

    script      0 0 OP_SUB
    script_hex  000094
    stack       0
    stack_hex                                {missing 00}
    results     success

    script      1 0 OP_SUB
    script_hex  510094
    stack       1
    stack_hex   01
    results     success

    script      127 0 127 1
    script_hex  017f00017f51
    stack       0x7f 0x 0x7f 1                {0x?}
    stack_hex   7f7f01                        {missing 00}
    results     success

### PR DESCRIPTION
**Pull request by ghost**
This fixes https://github.com/jb55/btcs/issues/41 and adds some more tests.

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -90,14 +90,10 @@ static int compile(int compile_options, int argc, const char *argv[])
...
 
-  stack_serialize(&tmp_stack, buf, bufsize, &compiled_len);
-
   if (compile_options & COMPILE_SHOW_STACK_HEX) {
 	  if (compile_options & COMPILE_SHOW_LABELS)
 		printf("stack_hex   ");
-	  for(i = 0; i < compiled_len; ++i)
-		  printf("%02x", buf[i]);
-	  printf("\n");
+		stack_print_hex(&tmp_stack);
```
Comment (jb55) [inline main.c]: I don't think this is correct? this should show the stack as a hexstring

File: op.c [hunk 2]
```diff
@@ -384,8 +384,13 @@ val_print(struct val val) {
   case VT_RAW: {
     u32 len;
     u8 *data = byte_pool_get(val.ind, &len);
-    printf("0x");
-    print_bytes(data, len, 0);
+    if (len == 0) {
+        printf("[]");
+    }
+    else {
+        printf("0x");
```
Comment (jb55) [inline op.c]: if `[]` means empty string than we should switch `0xdeadbeef` to `[deadbeff]`

File: script.c [hunk 3]
```diff
@@ -793,6 +793,30 @@ void script_print_vals(struct stack *stack) {
   putchar('\n');
 }
 
+void stack_print_hex(struct stack *stack) {
```
Comment (jb55) [inline script.c]: I don't think this implementation is correct. This should show the encoded value of the stack value. For the case of empty arrays, this should be nothing!

```
	  for(i = 0; i < compiled_len; ++i)
		  printf("%02x", buf[i]);
```

Think about this case:

`btcs 0 ripemd160` -> `9c1185a5c5e9fc54612808977ee8f548b2258d31`
`btcs @00 ripemd160` -> `c81b94933420221a7ac004a90242d8b1d3e5070d`

we wouldn't want an empty array to serialize to something other than empty or else hash operations would be wrong.

File: script_num.c [hunk 4]
```diff
@@ -34,19 +34,14 @@ sn_from_int(s64 n, struct num *sn) {
...
   const int neg = sn->val < 0;
   u64 absvalue = neg ? -(sn->val) : sn->val;
 
-  while(absvalue) {
+  do {
     *(p++) = absvalue & 0xff;
     assert((p - buf) <= bufsize);
     absvalue >>= 8;
-  }
+  } while(absvalue);
```
Comment (jb55) [inline script_num.c]: why was this changed, it looks wrong?

**Review by jb55 — CHANGES_REQUESTED** [review]

File: script.c (hunk 3 above)
Comment (jmacxx) [reply script.c]: Ok I will roll back that change.

File: script_num.c (hunk 4 above)
Comment (jmacxx) [reply script_num.c]: For the error case: `stack_hex                                {missing 00}` where 0 values are not displayed.  I changed the while loop to a do loop.  Further up in that function, when you have absvalue=0, it skips over the while loop and therefore the 0 is not serialized and not displayed.

File: op.c (hunk 2 above)
Comment (jmacxx) [reply op.c]: I'll roll it back.

File: main.c (hunk 1 above)
Comment (jmacxx) [reply main.c]: I'll roll it back.

File: script_num.c (hunk 4 above)
Comment (jmacxx) [reply script_num.c]: It fixes these failing tests:

    not ok 17 - test_OP_NOT_1
    #
    #    got      'script      1 OP_NOT script_hex  5191 stack       0 stack_hex    results     success', 
    #    expected 'script      1 OP_NOT script_hex  5191 stack       0 stack_hex   00 results     success'
    #
    not ok 18 - test_OP_NOT_2
    #
    #    got      'script      2 OP_NOT script_hex  5291 stack       0 stack_hex    results     success', 
    #    expected 'script      2 OP_NOT script_hex  5291 stack       0 stack_hex   00 results     success'
    #
    not ok 19 - test_OP_0NOTEQUAL0
    #
    #    got      'script      0 OP_0NOTEQUAL script_hex  0092 stack       0 stack_hex    results     success', 
    #    expected 'script      0 OP_0NOTEQUAL script_hex  0092 stack       0 stack_hex   00 results     success'
    #
    not ok 23 - test_BOOLAND
    #
    #    got      'script      0 0 OP_BOOLAND script_hex  00009a stack       0 stack_hex    results     success', 
    #    expected 'script      0 0 OP_BOOLAND script_hex  00009a stack       0 stack_hex   00 results     success'
    #

Comment (jb55) [reply script_num.c]: ok I'll take a closer look soon
